### PR TITLE
Improve handling out-of-RAM errors during Qdrant startup

### PR DESCRIPTION
**Pull request by ffuugoo**
Implemented basic out-of-RAM errors handling during Qdrant startup.

- add [`segment::utils::mem::Mem`] type that can check total and available RAM
  - it handles cgroups (v2 - tested, v1 - _supposedly_) memory limits, so it should work correctly
    inside a docker container
- add RAM checks to
  - [`open_simple_vector_storage`]
  - [`SimplePayloadStorage::open`]
  - [`GraphLinksRam::load_from_memory`]

__TODO:__
- [x] fix silly `cgroups_rs` crate errors (see failed CI runs)
  - possible solution is to replace `cgroups_rs` crate entirely ¯\\\_(ツ)\_/¯
- [x] add tests
- some possible improvements:
  - [ ] abstract `cgroups_mem::CgroupsMem` better
    - e.g., provide two implementations for Linux and all other OSes (with all no-op methods)
      to reduce the number of `#[cfg(...)]` annotations
  - [ ] add min/max required available memory limits during RAM checks
    - instead of flat `N%` _always_, which may be too low or too high depending on total memory
  - [ ] add transparent refresh wrapper that would call `Mem::refresh` automatically every `N` calls
    to `Mem::total_memory_bytes`/`Mem::available_memory_bytes` (to get rid of `if iter % 25 == 0 { .. }` blocks)
    and/or if more than `T` ms passed since last call to `Mem::refresh`
  - [ ] add a global shared `Mem` instance
    - (have to figure out how to synchronize access to it properly though)

[`segment::utils::mem::Mem`]: https://github.com/qdrant/qdrant/pull/1777/files#diff-2ae315f7c10d461299dcab65de97d722fd8c47923c45f690ec3b65535bfcf6e2
[`open_simple_vector_storage`]: https://github.com/qdrant/qdrant/pull/1777/files#diff-c83dc65ebae1d8e14fcea480596ea65f6e94f1762ed7009f69548215bb425cf8L55-L68
[`SimplePayloadStorage::open`]: https://github.com/qdrant/qdrant/pull/1777/files#diff-fbcfd74ef196bf5afc4b7343b8d12247887d699e138b9de71d88d8192ad56f9dL23-L29
[`GraphLinksRam::load_from_memory`]: https://github.com/qdrant/qdrant/pull/1777/files#diff-194df302a3af072ad57c7e465e75799f3a9cfe756d1e6158fa81df244e44810bL402-L414

### CODE REVIEW

**Comment by ffuugoo** [comment]
This was painful to debug. 🥲

**Comment by generall** [comment]
> This was painful to debug. smiling_face_with_tear

there-there

File: lib/segment/src/utils/mem.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        }
+
+        self.sysinfo.total_memory_bytes()
+    }
+
+    pub fn available_memory_bytes(&self) -> u64 {
+        #[cfg(target_os = "linux")]
+        if let Some(cgroups) = &self.cgroups {
+            if let Some(memory_limit_bytes) = cgroups.memory_limit_bytes() {
+                return memory_limit_bytes.saturating_sub(cgroups.used_memory_bytes());
```
Comment (timvisee) [inline lib/segment/src/utils/mem.rs:45]: Are in memory buffers and cache indeed excluded from the `used_memory_bytes()` value? I believe it is, just want to confirm.

**Review by timvisee — APPROVED** [review]
LGTM! Nice work on this. I like the cleanliness of the shell script.

I did dive into the cgroup documentation a bit to figure out what number of cgroup controllers are allowed. I've commented in the relevant code parts.

I agree that using a percentage may not be the best, as you've noted. But it does keep things simple.

It is extremely hard if not impossible to determine the actual _usable_ memory amount that you can request from the kernel. But I do think this is a good way to approximate it. I just hope this doesn't shoot us in the foot on a machine where everything would normally work perfectly fine.

File: lib/segment/src/utils/mem.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        pub fn used_memory_bytes(&self) -> u64 {
+            self.used_memory_bytes
+        }
+    }
+
+    fn get_current_process_memory_cgroup_path() -> procfs::ProcResult<Option<String>> {
+        let process = Process::myself()?;
+        let cgroups = process.cgroups()?;
+
+        for cgroup in cgroups {
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply lib/segment/src/utils/mem.rs:139]: Thanks for the review, @timvisee! I think it makes more sense to check for v2 first, as Docker (being the biggest slowpoke here) got support for (completely switched to?) cgroups v2, like, 2 years ago, and, generally, cgroups v2 are available for ~8 years now, so v1 is a legacy (and so, an afterthought) at this point.

File: lib/segment/src/utils/mem.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            // TODO: Can a process belong to multiple v2 cgroups!?
+            let is_v2_cgroup = cgroup.controllers.is_empty()
+                || cgroup
+                    .controllers
+                    .iter()
+                    .all(|controller| controller.is_empty());
+
+            // TODO: Can a process belong to multiple v1 cgroups, with some of these cgroups having the same controllers (e.g., memory)!?
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply lib/segment/src/utils/mem.rs:147]: Kudos for verifying this! 🙏

File: lib/segment/src/utils/mem.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (timvisee) [reply lib/segment/src/utils/mem.rs:139]: Makes sense!

I figured that the v1 check may be a bit more reliable because we test for a specific controller rather than for it being empty.

But on second thought I agree, check v2 first.

[the remark this answers is not in the document]
Comment (ffuugoo) [reply lib/segment/src/utils/mem.rs:139]: I figured v1 _have to_ have _some_ controller. (I may be wrong, though. I focused on and tested v2.)

**Review by generall — CHANGES_REQUESTED** [review]
Sorry, I am not convinced those changes are not going to be a foot-gun.
Here are some problems I see:

- memory limit is hard-coded and enabled by default. That _will be_ a breaking change for someone
- OOM handling if implemented in the same way as OOD, which is not correct. E.g. if we try to create a new collection: OOD will be handled and OOM will not. There is a DB lock API we should use.

Overall, I propose to rollback all changes except for cgroups function

**Comment by generall** [comment]
Will try to propose an alternative shortly

**Comment by ffuugoo** [comment]
> - memory limit is hard-coded and enabled by default. That _will be_ a breaking change for someone
> 
> - OOM handling if implemented in the same way as OOD, which is not correct. E.g. if we try to create a new collection: OOD will be handled and OOM will not. There is a DB lock API we should use.

I feel like this can be improved iteratively:
- add a config to enable OOM handling and configure limits
- add additional `OperationError::WouldOOM` error variant and lock DB if `segment`/`collection` crate returns it

Though, I'm absolutely open to better suggestions, cause I don't have a better overall approach (i.e., periodically check for available memory during load and abort) to OOM handling.

**Comment by generall** [comment]
Replaced periodic checks in the in-mem vector storage and HNSW graph loading to `try_reserve`, so it might also be caught in runtime. 
Unfortunately, it won't work for the payload storage/index, as there are lot of small allocations and some structures there doesn't support `try_reserve`.

I have one idea to try, stay tuned

**Comment by generall** [comment]
here it is https://github.com/qdrant/qdrant/pull/1809

**Review by generall — APPROVED** [review]
